bin/node/executor/src/lib.rs

```diff
@@ -25,5 +25,6 @@ use sc_executor::native_executor_instance;
 native_executor_instance!(
 	pub Executor,
 	node_runtime::api::dispatch,
-	node_runtime::native_version
+	node_runtime::native_version,
+	sp_io::benchmarking::HostFunctions,
```

Comment:
It's still exposed, but only in `node`, isn't it? Why not use `Capabilities` and `Externalities` extensions for this?

Reply:
This is not an extension, these are the actual host functions that will be exposed by the wasm executor.

So, what you want will not work.

Reply:
I'm not sure I get how these functions here differ from extensions. To best of my knowledge, extensions are always exposed, but are not always callable. This here will also always expose the bechmarking functions and make them always callable, but only when you are using `node/executor`. So it means benchmarking functions are going to be available even when you import blocks (running `node` obviously). To me it doesn't sound like it's superior to extensions then unless we actually want to measure real import time / extrinsic dispatch. but even for this I'm sure we could come up with API that is deterministic (but just have non-deterministic side effects, like printing timing log lines).

Reply:
This operates on a level below `Extensions`. Let's explain at the example of offchain runtime interfaces. We always expose the offchain runtime interfaces to the runtime, this is required because the same runtime will be used to execute the offchain worker. However, when you call an offchain function from inside the block production/import it will fail because we did not give it the extension. However, that means that the function is callable.

Now to this change here. A runtime can still call these benchmarking functions, that is right. However, by default it will fail at instantiation of the runtime, because these functions are not given to the wasm-executor and thus, they are not exposed to the runtime. If you try to do this, you will get an error that complains about missing functions bla. 

However, as we need the benchmarking functionality in the Substrate node, we instruct the `Executor` to also expose the benchmarking functions. So, it will not fail at instantiation of the runtime.

Reply:
Yup, but it means that it's still possible to build a runtime that uses this functions, compile it and if it's going to be run on the correct executor (default `node/executor` here) it will just work fine, exposing non-deterministic functions to the state transition function.

To me this is going to set a precedence of having non-deterministic functions callable, with only a warning in the docs. I was hoping we can do better than this to protect runtime devs, but it seems I'm in minority, and since we're not creating children's toys here, I won't oppose.

Reply:
If you tell me how? :D 

It will run on the Substrate node, but that is anyway some special node that we use for benchmarking etc. When you create your own node, you should not expose this functionality to the executor! Any executor created by `native_executor_instance` will fail to instantiate your runtime when you try to call one of these functions.

We will also improve here in the node with a benchmarking feature or similar. However, this requires some changes to `wasm-builder` and `sp-api` to make it work correctly.